README.md
### Description
This is an R package to convert R graphics to Flash file (SWF format).
The conversion can be done in two different ways:

1. This package contains a SWF device in R, so you can create SWF file
directly using plotting functions like `plot()` and `lines()`.
The convention is that every time you call a high-level plotting function,
e.g. `plot()`, the movie will create a new frame and draw the following
shapes in this frame. In this way, you can create some animation by
calling a series of `plot()` functions.
2. This package also includes several functions to convert graphics of
other formats into SWF format. For example, `svg2swf()` to convert from
SVG to SWF, and `image2swf()` to convert PNG and JPG images into a single
SWF file.

### Installation
`R2SWF` depends on the following libraries:

- zlib         [http://www.zlib.net/](http://www.zlib.net/)
- libpng       [http://www.libpng.org/](http://www.libpng.org/)
- freetype     [http://www.freetype.org/](http://www.freetype.org/)
- libming      [http://www.libming.org/](http://www.libming.org/)

The source code of `libming` is included in `R2SWF`, and you need
to install the other three by yourself.

For Debian/Ubuntu users, the command to install dependent libraries is

```
sudo apt-get install zlib1g-dev libpng12-dev libfreetype6-dev
```

For rpm based systems (e.g. Fedora), try to run

```
sudo yum install zlib-devel libpng-devel freetype-devel
```

### Examples
In the first example, we first create 20 images using `png()` function,
and then convert them into a single SWF file `R2SWF-ex1.swf`.

```r
## Creating png files
png("image-png-%03d.png", 480, 300)
x = seq(0, 2 * pi, length.out = 20)
cols = rainbow(20)
for(i in 1:20) plot(x[i], sin(x[i]), xlim = c(0, 2 * pi), ylim = c(-1, 1),
                    col = cols[i], pch = 16, cex = 2, main = "PNG => SWF")
dev.off()

## Obtain the filenames
pngfiles = sprintf("image-png-%03d.png", 1:20)

## Convert to SWF
image2swf(pngfiles, "R2SWF-ex1.swf", interval = 0.3)

```

Using `svg2swf` is pretty similar, except that the output animation contains vector graphics.

```r
## Do similar things as above
svg("image-svg-%03d.svg", 8, 5)
x = seq(0, 2 * pi, length.out = 20)
cols = rainbow(20)
for(i in 1:20) plot(x[i], sin(x[i]), xlim = c(0, 2 * pi), ylim = c(-1, 1),
                    col = cols[i], pch = 16, cex = 2, main = "SVG => SWF")
dev.off()
svgfiles = sprintf("image-svg-%03d.svg", 1:20)

## Convert to SWF
svg2swf(svgfiles, "R2SWF-ex2.swf", interval = 0.3)

```

The third example shows how to use the SWF device to create (rather than converting) SWF file directly.

```r
swf("R2SWF-ex3.swf")
set.seed(123)
x = rnorm(5)
y = rnorm(5)
for(i in 1:100) {
    plot(x <- x + 0.1 * rnorm(5), y <- y + 0.1 * rnorm(5),
         xlim = c(-3, 3), ylim = c(-3, 3), col = "steelblue",
         pch = 16, cex = 2, xlab = "x", ylab = "y")
    title("Brownian Motion")
}
dev.off()
```

In general, when using the SWF device, high-level plotting functions (e.g. `plot()`)
will advance the movie by one frame, and low-level functions (`lines()`, `text()`, etc.)
are effective only to the current frame.




NEWS
          CHANGES IN R2SWF VERSION 0.9

BUG FIXES

    o Fix a bug in setting the dimension of the SWF file that occurs in
      image2swf()
    o Eliminate R CMD check warnings





          CHANGES IN R2SWF VERSION 0.8-2

BUG FIXES

    o Fix a misuse of abs() in the C code, thanks to Prof. Ripley





          CHANGES IN R2SWF VERSION 0.8-1

BUG FIXES

    o Fix a memory bug detected by valgrind, thanks to Prof. Ripley





          CHANGES IN R2SWF VERSION 0.8

NEW FEATURES

    o SWF graphics device now supports the hadj parameter


DEPRECATED AND DEFUNCT

    o Functions list.fonts() and add.fonts() are removed
    o Functions font.paths(), font.families(), font.files() and
      font.add() are moved to the sysfonts package which the
      current version of R2SWF depends on





          CHANGES IN R2SWF VERSION 0.7-1

BUG FIXES

    o Fix a bug in detecting the OS type





          CHANGES IN R2SWF VERSION 0.7

NEW FEATURES

    o Add function font.files() to list available font files


BUG FIXES

    o Fix swf2html() to make it work properly with knitr hook hook_r2swf()
    o Rename list.fonts() to font.families(), and add.fonts() to font.add().
      Old function names are deprecated but kept for compatibility.





          CHANGES IN R2SWF VERSION 0.6-1

BUG FIXES

    o Remove configure.win so it should be built on both architectures
      on Windows
    o Fix loading problem on R 2.15.3





          CHANGES IN R2SWF VERSION 0.6

BUG FIXES

    o Clarify the copyrights
    o Remove the source code of some external libraries





          CHANGES IN R2SWF VERSION 0.5

NEW FEATURES

    o A SWF device for R through swf()


BUG FIXES

    o Take care of absolute and relative paths of files
    o Remove "outdir" argument from dev2swf()





          CHANGES IN R2SWF VERSION 0.4-1

BUG FIXES

    o Update zlib to 1.2.8
    o Update libpng to 1.6.2
    o Add freetype 2.5.0
    o Update libming to 0.4.5
    o Fix several memory leak problems
    o Use absolute paths for dev2swf(), swf2html() etc.
    o HTML file created by swf2html() will have a copy of the SWF file
      in the same directory





          CHANGES IN R2SWF VERSION 0.4

NEW FEATURES

    o A new function file2swf() as a wrapper of image2swf() and
      svg2swf()

    o swf2html() uses a simple <embed> tag instead of the long ugly old
      <object> tag to write the Flash file in HTML; it also gained a new
      argument 'fragment' to decide whether to return an HTML fragment or
      a complete HTML document





          CHANGES IN R2SWF VERSION 0.3-1

NEW FEATURES

    o Windows version of R supports svg() since 2.14.0.
      Now svg2swf() also accepts SVG files created by svg().
    
    o dev2swf() now supports svg() device





          CHANGES IN R2SWF VERSION 0.3-0

NEW FEATURES

    o Allow users to set the background color of the output SWF file

    o Re-design the arguments list of image2swf() and svg2swf()
      for ease of use

BUG FIXES

    o (Internal) Use a more appropriate way to insert frames in a SWF file





          CHANGES IN R2SWF VERSION 0.2-2

BUG FIXES

    o Fix Makefile errors





          CHANGES IN R2SWF VERSION 0.2-1

BUG FIXES

    o Fix compilation errors in Windows





          CHANGES IN R2SWF VERSION 0.2-0

NEW FEATURES

    o svg2swf() to convert SVG files to a Flash animation





          CHANGES IN R2SWF VERSION 0.1-0

NEW FEATURES

    o image2swf() and dev2swf() to convert png/jpeg files to a Flash animation

    o swf2html() to embed a Flash file into an HTML page


R/svg2swf.R
# Currently this function could only parse svg files created by the cairo
# graphics library, typically from svg() in the grDevices package (R >= 2.14.0
# required for Windows OS), and CairoSVG() in the Cairo package.
parseSVG = function(file.name) {
  xmlParse = XML::xmlParse;
  newXMLNamespace = XML::newXMLNamespace;
  xmlRoot = XML::xmlRoot;
  getNodeSet = XML::getNodeSet;
  xmlAttrs = XML::xmlAttrs;
  xmlParent = XML::xmlParent;
  xmlName = XML::xmlName;
  xmlChildren = XML::xmlChildren;
  xmlSApply = XML::xmlSApply;

  svgFile = xmlParse(file.name);
  # Don't forget the name space!
  newXMLNamespace(xmlRoot(svgFile), "http://www.w3.org/2000/svg", "svg");
  # Find the first <g> child of <svg>
  pathRoot = getNodeSet(svgFile, "/svg:svg/svg:g");
  if(!length(pathRoot)) stop(sprintf("Failed in parsing file '%s'", file.name));
  pathRoot = pathRoot[[1]];

  # Default style for a <path> node
  defaultStyle = c("stroke" = "none",
                   "stroke-width" = "1",
                   "stroke-linecap" = "butt",
                   "stroke-linejoin" = "miter",
                   "stroke-miterlimit" = "4",
                   "stroke-opacity" = "1",
                   "fill" = "rgb(0%,0%,0%)",
                   "fill-rule" = "nonzero",
                   "fill-opacity" = "1");

  # Handle <path> style in named vector
  parseStyle = function(style)
  {
    if(is.null(style)) return(NULL);
    s = unlist(strsplit(style, ";"));
    val = strsplit(s, ":");
    result = sapply(val, function(x) x[2]);
    names(result) = sapply(val, function(x) x[1]);
    return(result);
  }
  # Update the attributes in "old" style with the values in "new"
  # "old" must contain "new"
  updateStyle = function(old, new)
  {
    if(is.null(new)) return(old);
    result = old;
    result[names(new)] = new;
    return(result);
  }
  # Iteratively update the style from parent nodes
  updateStyleUpward = function(node)
  {
    style = xmlAttrs(node)["style"];
    if(is.na(style)) style = NULL;
    style = parseStyle(style);
    style = updateStyle(defaultStyle, style);
    parentNode = xmlParent(node);
    # Recursively search the parent
    while(!is.null(parentNode))
    {
      parentStyle = xmlAttrs(parentNode)["style"];
      if(is.null(parentStyle) || is.na(parentStyle)) parentStyle = NULL;
      parentStyle = parseStyle(parentStyle);
      style = updateStyle(style, parentStyle);
      parentNode = xmlParent(parentNode);
    }
    return(style);
  }
  # Parse <path> and <use> nodes into structured lists
  #
  # <path style="" d="">   =====>   style=..., d=..., x=0, y=0
  #
  # <use xlink:href="#glyph0-0" x="63.046875" y="385.921875"/>
  # =====>
  # style=..., d=..., x=63.046875, y=385.921875
  #
  parseNode = function(node)
  {
    if(xmlName(node) == "use")
    {
      attrs = xmlAttrs(node);
      refID = sub("#", "", attrs["href"]);
      refPathNode = getNodeSet(svgFile, sprintf("//*[@id='%s']/svg:path", refID))[[1]];
      style = updateStyleUpward(refPathNode);
      style = updateStyle(style, updateStyleUpward(node));
      d = xmlAttrs(refPathNode)["d"];
      x = xmlAttrs(node)["x"];
      y = xmlAttrs(node)["y"];
    } else if(xmlName(node) == "path") {
      style = updateStyleUpward(node);
      d = xmlAttrs(node)["d"];
      x = y = 0;
    } else return(NULL);
    xy = as.numeric(c(x, y));
    names(d) = NULL;
    names(xy) = NULL;
    return(list(style = style, d = d, xy = xy));
  }
  # Flatten nodes
  # <g>
  #   <use />
  #   <use />
  #   <use />
  # </g>
  #
  # =====>
  #
  # <use />
  # <use />
  # <use />
  expandNode = function(node)
  {
    children = xmlChildren(node);
    res = if(!length(children)) node else children;
    return(res);
  }
  nodes = unlist(xmlSApply(pathRoot, expandNode));
  names(nodes) = NULL;
  paths = lapply(nodes, parseNode);
  path.is.null = sapply(paths, is.null);
  paths[path.is.null] = NULL;
  if(!length(paths)) stop("Unknown child node of '/svg/g'");
  return(paths);
}


#' Convert a sequence of SVG files to SWF file
#'
#' Given the file names of a sequence of SVG files, this function could
#' convert them into a Flash file (.swf).
#'
#' This function uses the XML package in R and a subset of librsvg
#' (\url{http://librsvg.sourceforge.net/}) to parse the SVG file, and
#' uses the Ming library (\url{http://www.libming.org/}) to
#' implement the conversion. Currently this function supports SVG files
#' created by \code{\link[grDevices]{svg}()} in the \pkg{grDevices}
#' package, and \code{\link[Cairo]{CairoSVG}()} in the
#' \pkg{Cairo} package.
#' @param input the file names of the SVG files to be converted
#' @param output the name of the output SWF file
#' @param bgColor background color of the output SWF file
#' @param interval the time interval (in seconds) between animation frames
#' @return The name of the generated SWF file if successful.
#' @export
#' @author Yixuan Qiu <\email{yixuan.qiu@@cos.name}>
#' @examples \dontrun{
#' if(capabilities("cairo")) {
#'   olddir = setwd(tempdir())
#'   svg("Rplot%03d.svg", onefile = FALSE)
#'   set.seed(123)
#'   x = rnorm(5)
#'   y = rnorm(5)
#'   for(i in 1:100) {
#'       plot(x <- x + 0.1 * rnorm(5), y <- y + 0.1 * rnorm(5),
#'            xlim = c(-3, 3), ylim = c(-3, 3), col = "steelblue",
#'            pch = 16, cex = 2, xlab = "x", ylab = "y")
#'   }
#'   dev.off()
#'   output = svg2swf(sprintf("Rplot%03d.svg", 1:100), interval = 0.1)
#'   swf2html(output)
#'   setwd(olddir)
#' }
#' }
#'
svg2swf = function(input, output = "movie.swf", bgColor = "white",
                   interval = 1) {
  # Use XML package
  # if(!require(XML))
  #     stop("svg2swf() requires XML package");
  
  xmlParse = XML::xmlParse;
  xmlAttrs = XML::xmlAttrs;
  xmlRoot = XML::xmlRoot;
  
  if(!is.character(input))
    stop("'input' must be a character vector naming the input SVG files");

  bg = col2rgb(bgColor, alpha = FALSE);
  bg = as.integer(bg);

  if(!all(file.exists(input))) stop("one or more input files do not exist");
  
  filesData = lapply(input, parseSVG);
  firstFile = xmlParse(input[1]);
  size = xmlAttrs(xmlRoot(firstFile))["viewBox"];
  size = as.numeric(unlist(strsplit(size, " ")));

  outfile = normalizePath(output, mustWork = FALSE);
  .Call("svg2swf", filesData, outfile, size,
        bg, as.numeric(interval), PACKAGE = "R2SWF");

  message("SWF file created at ", outfile);
  invisible(output);
}
